TiFlash meet bus error and crash in some specific env

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
CPU: Kunpeng-920(arm)
OS: 
```
Kylin Linux Advanced Server release V10 (Sword)
Linux localhost.localdomain 4.19.90-24.4.v2101.ky10.aarch64 #1 SMP Mon May 24 14:45:37 CST 2021 aarch64 aarch64 aarch64 GNU/Linux
```
log
```
[2025/05/14 01:46:18.005 +08:00] [ERROR] [BaseDaemon.cpp:366] [########################################] [source=BaseDaemon] [thread_id=780]
[2025/05/14 01:46:18.005 +08:00] [ERROR] [BaseDaemon.cpp:367] ["(from thread 632) Received signal Bus error(7)."] [source=BaseDaemon] [thread_id=780]
[2025/05/14 01:46:18.005 +08:00] [ERROR] [BaseDaemon.cpp:426] ["Invalid address alignment."] [source=BaseDaemon] [thread_id=780]
[2025/05/14 01:46:22.843 +08:00] [ERROR] [BaseDaemon.cpp:559] ["\n  0xaaad99324274\tfaultSignalHandler(int, siginfo_t*, void*) [tiflash+121258612]\n                \tlibs/libdaemon/src/BaseDaemon.cpp:210\n  0xfffd2ce107c0\t<unknown symbol> [linux-vdso.so.1+1984]\n  0xff991e4fc4a8\t<unknown symbol>"] [source=BaseDaemon] [thread_id=780]
```
After some debug, we found the crash point is in
https://github.com/pingcap/tiflash/blob/5c59a514c5fec73b26e6773b57b968f2f59afb7f/dbms/src/Interpreters/JoinPartition.cpp#L1137

<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)

### 3. What did you see instead (Required)

### 4. What is your TiFlash version? (Required)

--- a/dbms/src/Common/tests/gtest_arena.cpp
+++ b/dbms/src/Common/tests/gtest_arena.cpp
@@ -1,4 +1,4 @@
-// Copyright 2023 PingCAP, Inc.
+// Copyright 2025 PingCAP, Inc.
 //
 // Licensed under the Apache License, Version 2.0 (the "License");
 // you may not use this file except in compliance with the License.

--- a/dbms/src/Common/tests/gtest_arena.cpp
+++ b/dbms/src/Common/tests/gtest_arena.cpp
@@ -26,15 +26,15 @@ TEST(ArenaTest, AlignedAlloc)
     auto pool = DB::Arena(1024);
     auto * p = pool.alloc(3);
     p = pool.alloc(2);
-    // the first allocation is not aligned
-    ASSERT_NE(reinterpret_cast<uintptr_t>(p) % 16, 0);
-    // the second allocation is aligned
     auto align = alignof(CachedColumnInfo);
+    // the first allocation is not aligned
+    ASSERT_NE(reinterpret_cast<uintptr_t>(p) % align, 0);
+    // the second allocation is aligned
     p = pool.alignedAlloc(sizeof(CachedColumnInfo), align);
     ASSERT_EQ(reinterpret_cast<uintptr_t>(p) % align, 0);
     auto * cached_column_info = reinterpret_cast<CachedColumnInfo *>(p);
     new (cached_column_info) CachedColumnInfo(nullptr);
-    // double check the alignment
+    // double check the alignment, will raise bus error if not aligned in some platforms
     cached_column_info->mu.lock();
     cached_column_info->mu.unlock();
 }

--- a/dbms/src/Common/tests/gtest_arena.cpp
+++ b/dbms/src/Common/tests/gtest_arena.cpp
@@ -1,0 +1,32 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Common/Arena.h>
+#include <Interpreters/JoinHashMap.h>
+#include <gtest/gtest.h>
+
+namespace DB::tests
+{
+TEST(ArenaTest, AlignedAlloc)
+{
+    auto pool = DB::Arena(1024);
+    auto * p = pool.alloc(3);
+    p = pool.alloc(2);
+    // the first allocation is not aligned
+    ASSERT_NE(reinterpret_cast<uintptr_t>(p) % 16, 0);
+    // the second allocation is aligned
+    p = pool.alignedAlloc(sizeof(CachedColumnInfo), 16);
+    ASSERT_EQ(reinterpret_cast<uintptr_t>(p) % 16, 0);
+}
+} // namespace DB::tests

--- a/dbms/src/Common/tests/gtest_arena.cpp
+++ b/dbms/src/Common/tests/gtest_arena.cpp
@@ -16,6 +16,9 @@
 #include <Interpreters/JoinHashMap.h>
 #include <gtest/gtest.h>
 
+#include <mutex>
+#include <type_traits>
+
 namespace DB::tests
 {
 TEST(ArenaTest, AlignedAlloc)
@@ -26,7 +29,13 @@ TEST(ArenaTest, AlignedAlloc)
     // the first allocation is not aligned
     ASSERT_NE(reinterpret_cast<uintptr_t>(p) % 16, 0);
     // the second allocation is aligned
-    p = pool.alignedAlloc(sizeof(CachedColumnInfo), 16);
-    ASSERT_EQ(reinterpret_cast<uintptr_t>(p) % 16, 0);
+    auto align = alignof(CachedColumnInfo);
+    p = pool.alignedAlloc(sizeof(CachedColumnInfo), align);
+    ASSERT_EQ(reinterpret_cast<uintptr_t>(p) % align, 0);
+    auto * cached_column_info = reinterpret_cast<CachedColumnInfo *>(p);
+    new (cached_column_info) CachedColumnInfo(nullptr);
+    // double check the alignment
+    cached_column_info->mu.lock();
+    cached_column_info->mu.unlock();
 }
 } // namespace DB::tests

--- a/dbms/src/Interpreters/JoinPartition.cpp
+++ b/dbms/src/Interpreters/JoinPartition.cpp
@@ -45,7 +45,7 @@ void insertRowToList(JoinArenaPool & pool, List * list, Elem * elem, size_t cach
         if unlikely (cache_columns_threshold == list->list_length)
         {
             auto * cached_column_info
-                = reinterpret_cast<CachedColumnInfo *>(pool.arena.alloc(sizeof(CachedColumnInfo)));
+                = reinterpret_cast<CachedColumnInfo *>(pool.arena.alignedAlloc(sizeof(CachedColumnInfo), 16));
             new (cached_column_info) CachedColumnInfo(list->next);
             pool.cached_column_infos.push_back(cached_column_info);
             list->cached_column_info = cached_column_info;

--- a/dbms/src/Interpreters/JoinPartition.cpp
+++ b/dbms/src/Interpreters/JoinPartition.cpp
@@ -20,7 +20,6 @@
 #include <Interpreters/ProbeProcessInfo.h>
 
 #include <ext/scope_guard.h>
-#include <type_traits>
 
 namespace DB
 {

--- a/dbms/src/Interpreters/JoinPartition.cpp
+++ b/dbms/src/Interpreters/JoinPartition.cpp
@@ -20,6 +20,7 @@
 #include <Interpreters/ProbeProcessInfo.h>
 
 #include <ext/scope_guard.h>
+#include <type_traits>
 
 namespace DB
 {
@@ -44,8 +45,8 @@ void insertRowToList(JoinArenaPool & pool, List * list, Elem * elem, size_t cach
     {
         if unlikely (cache_columns_threshold == list->list_length)
         {
-            auto * cached_column_info
-                = reinterpret_cast<CachedColumnInfo *>(pool.arena.alignedAlloc(sizeof(CachedColumnInfo), 16));
+            auto * cached_column_info = reinterpret_cast<CachedColumnInfo *>(
+                pool.arena.alignedAlloc(sizeof(CachedColumnInfo), alignof(CachedColumnInfo)));
             new (cached_column_info) CachedColumnInfo(list->next);
             pool.cached_column_infos.push_back(cached_column_info);
             list->cached_column_info = cached_column_info;

